File: floating_point_experiments/float_comparison.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char* argv[])
{
  float f = 0.1f, sum = 0;
  int max = atoi(argv[1]);
   
  printf("   %25s %25s %25s\n", "sum", "float product", "double product");
  for (int i = 0; i < max; ++i)
  {
    float float_product = f * i;
    double double_product = ((double)f) * i;

    printf("%2d) %1.23f %1.23f %1.23f\n", i, sum, float_product, double_product);

    sum += f;
  }
}

File: floating_point_experiments/float_comparison.py
#! /usr/bin/env python

import sys

s = 0.1
print("    {:25s} {:25s} {:25s}".format("sum", "product", "difference"))
for i in range(1, int(sys.argv[1])):
  m = (0.1 * float(i))
  d = abs(m - s)
  print("{:2d}) {:1.23f} {:1.23f} {:1.23f}".format(i,s,m,d))
  s = s + 0.1

